Mathlib/Algebra/Ring/Ext.lean

```diff
@@ -0,0 +1,477 @@
+/-
+Copyright (c) 2023 Raghuram Sundararajan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Raghuram Sundararajan
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Ext
+-- import Mathlib
+
+/-!
+# Extensionality lemmas for rings and similar structures
+
+In this file we prove extensionality lemmas for the ring-like structures defined in
+`Algebra/Ring/Defs`, ranging from `NonUnitalNonAssocSemiring` to `CommRing`. These extensionality
+lemmas take the form of asserting that two algebraic structures on a type are equal whenever the
+addition and multiplication defined by them are both the same.
+
+## Implementation details
+
+We follow `Algebra/Group/Ext` in using the term `(letI := i; HMul.hMul : R → R → R)` to refer to the
+multiplcation specified by a typeclass instance `i` on a type `R` (and similarly for addition). We
+abbreviate these using some local notations.
+
+Since `Algebra/Group/Ext` proved several injectivity lemmas, we do so as well — even if sometimes we
+don't need them to prove extensionality.
+
+## Tags
+semiring, ring, extensionality
+-/
+
+local macro:65 lhs:term:65 "+[" type:term ", " inst:term "]" rhs:term:66 : term =>
+  `(term| (letI := $inst; ($lhs : $type) + ($rhs : $type) : $type))
+local macro:70 lhs:term:70 "*[" type:term ", " inst:term "]" rhs:term:71 : term =>
+  `(term| (letI := $inst; ($lhs : $type) * ($rhs : $type) : $type))
+
+universe u
+
+variable {R : Type u}
+
+/-! ### Distrib -/
+namespace Distrib
+
+@[ext] theorem ext ⦃inst₁ inst₂ : Distrib R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ := by
+  -- Split into `add` and `mul` functions and properties.
+  rcases inst₁ with @⟨⟨⟩, ⟨⟩⟩
+  rcases inst₂ with @⟨⟨⟩, ⟨⟩⟩
+  -- Prove equality of parts using function extensionality.
+  congr <;> (apply funext₂; assumption)
+
+theorem ext_iff (inst₁ inst₂ : Distrib R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
```

Comment:
We don't have these `ext_iff` versions for `Monoid` and friends (and I suspect they're not actually useful), so I'd be inclined to skip them for now.

Reply:
I added them because the `ext` attribute generates them for `class` definitions (probably `structure` and `inductive` as well).

Searching through Mathlib does turn up several uses of `ext_iff` lemmas, both as `ext_iff.1`/`ext_iff.mp` (which I guess is essentially a convenient way to avoid having to incant the right `congr` or `injection`s) and in `simp`s, although I don't know if these ones are going to be of any use.

I guess there's no harm in keeping these, even if we don't keep writing them?

One thing I did notice is that for `ext_iff` to be usable as `ext_iff.1`/`ext_iff.mp`, it should only take instance or implicit arguments (although this is not the behaviour of `@[ext]`-generated `ext_iff` lemmas).
If these `ext_iff` are kept, they could be changed to do the same (or not).

Reply:
Now that they are written, I don't see much harm in keeping the `ext_iff`s. In this case the autogeneration behaviour of `@[ext]` is probably the wrong thing to do: as a rule of thumb, arguments that appear on both sides of an `iff` can be safely left implicit.

Here's another way to write the proof. Using `rintro rfl` means both sides are immediately equal, so no `congr` is needed. But either way is fine really.
```suggestion
theorem ext_iff (inst₁ inst₂ : Distrib R) :
    inst₁ = inst₂ ↔
      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
  ⟨by rintro rfl; constructor <;> (intros; rfl), And.elim (ext · ·)⟩
```